language/move-vm/types/src/gas_schedule.rs

```diff
@@ -6,21 +6,117 @@
 //! It is important to note that the cost schedule defined in this file does not track hashing
 //! operations or other native operations; the cost of each native operation will be returned by the
 //! native function itself.
-use libra_types::transaction::MAX_TRANSACTION_SIZE_IN_BYTES;
+use libra_types::{
+    transaction::MAX_TRANSACTION_SIZE_IN_BYTES,
+    vm_error::{StatusCode, VMStatus},
+};
 use mirai_annotations::*;
 use move_core_types::gas_schedule::{
     words_in, AbstractMemorySize, CostTable, GasAlgebra, GasCarrier, GasConstants, GasCost,
     GasUnits,
 };
 use vm::{
+    errors::VMResult,
     file_format::{
         Bytecode, ConstantPoolIndex, FieldHandleIndex, FieldInstantiationIndex,
         FunctionHandleIndex, FunctionInstantiationIndex, StructDefInstantiationIndex,
         StructDefinitionIndex, NUMBER_OF_NATIVE_FUNCTIONS,
     },
-    file_format_common::instruction_key,
+    file_format_common::{instruction_key, Opcodes},
 };
 
+/// The Move VM implementation for gas charging.
```

Comment:
this is the gas Api to be provided to the Move VM.
It is a concrete implementation and it's forced to clients.
It's the most reasonable way I could see this working for now.
I think this is much better than any trait story

Reply:
I think the concrete implementation is good at least for now and probably for v1, but I don't see the downsides of a trait story either. Could you elaborate? My gut feeling is that a trait story might help us move certain concepts out of the Move VM and make it possible to enable/disable gas metering via monomorphization.